### PR TITLE
Implementation_of_Error_Metric_for_Rest_Actions

### LINKED ISSUE #4401: Implementation of  Error Metric for Rest Actions
*Is your feature request related to a problem? Please describe.*
Whenever 4xx or 5xx errors occur for any API call, we don’t have way to know what number of errors caused by which API.  
This feature is to provide customer a sort of metrics for API-wise 4xx and 5xx errors. This helps to monitor the system behaviour for different APIs as well.

*Describe the solution you'd like*
Whenever an API throws 4xx or 5xx, we’ll keep 4xx and 5xx error count in-memory for that API. We are going to use an existing API (/_nodes/stats) to return that count metric. This API is already being used to retrieve the statistics. We are going to add a new matrix in the response named “rest_actions” which will contain number of times 4xx and 5xx error occurred for an API. As of now we’ll capture the error count for indexing, search API and other APIs error count is combined.
New metric to be introduced : “rest_actions”
For API call : GET /_nodes/stats/rest_actions
Response will look like as follows...

{
    "_nodes": {
        "total": 2,
        "successful": 2,
        "failed": 0
    },
    "cluster_name": "docker-cluster",
    "nodes": {
        "hD-meqDVTTm7fcqCP2F7xQ": {
            "rest_actions": {
                "errors": {
                    "indexing": {
                        "5xx": 10,
                        "4xx": 2
                    },
                    "searching": {
                        "5xx": 1,
                        "4xx": 0
                    },
                    "others": {
                        "5xx": 20,
                        "4xx": 4
                    }
                }
            }
        },
        "pHa-gp3ES_K4PfETNUBqzA": {
            "rest_actions": {
                "errors": {
                    "indexing": {
                        "5xx": 10,
                        "4xx": 2
                    },
                    "searching": {
                        "5xx": 1,
                        "4xx": 0
                    },
                    "others": {
                        "5xx": 20,
                        "4xx": 4
                    }
                }
            }
        }
    }
}

This approach is considered is good option for below reasons...
Easy and quick to implement
Can be extended further if more info required easily
Independent to node type

*Describe alternatives you've considered*
Another approach was to use PA plugin. We keep the error count same way as above approach but PA plugin would collect those matrix instead of exposing via API. PA collects the error count every 5 sec and store them in matricDB files.
This approach is discarded for following reasons...
PA is a separate component which runs on every node, so extra work to be done in order to collect matrix every 5 sec
It store the matrix in storage hence additional storage is required
Have some limitation as it does’t run on T2 nodes.

One more  alternate approach was to use (/_nodes/usage) to return the response but this API is specific to get usage for ex, it returns number of time a rest action has been called.  

"rest_actions": {
    "nodes_usage_action": 1,
    "get_indices_action": 4,
    "nodes_stats_action": 2,
    "main_action": 2,
    "PerformanceAnalyzerClusterConfigAction": 3
}

Therefore, the error count doesn’t fall under this category. And chose (stats) API.

*Additional context*

### PR DESCRIPTION
**Pull request by vzez**
### Description
Complete implementation of issue: #4401 

### Issues Resolved
#4401 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodesStatsRequest.java [hunk 1]
```diff
@@ -79,6 +81,7 @@ public NodesStatsRequest(StreamInput in) throws IOException {
             optionallyAddMetric(in.readBoolean(), Metric.ADAPTIVE_SELECTION.metricName());
         } else {
             requestedMetrics.addAll(in.readStringList());
+            restActionsFilters.addAll(in.readStringList());
```
Comment (mgodwan) [inline server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodesStatsRequest.java:84]: What happens if no filters are provided?

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/RestActionsStats.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Rest actions responses stats
+ *
+ * @opensearch.internal
+ */
+public class RestActionsStats implements Writeable, ToXContentFragment {
+    private Map<String, Map<Integer, Integer>> restActionsStatusCount;
```
Comment (mgodwan) [inline server/src/main/java/org/opensearch/action/admin/cluster/node/stats/RestActionsStats.java:29]: Can we use Long here?

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/RestActionsStats.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    private Map<Integer, Integer> restActionsStatusTotalCount;
```
Comment (mgodwan) [inline server/src/main/java/org/opensearch/action/admin/cluster/node/stats/RestActionsStats.java:30]: 1. Can you add documentation for the fields here?
2. Is the key supposed to denote the status code?

**Comment by mgodwan** [comment]
Could you help with the sample request and response structure for this?

File: server/src/main/java/org/opensearch/rest/action/admin/cluster/RestNodesStatsAction.java [hunk 4]
```diff
@@ -185,6 +195,19 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
                     )
                 );
             }
+            if (metrics.contains("rest_actions") && request.hasParam("rest_actions_filters")) {
+                nodesStatsRequest.addRestActionsFilters(request.paramAsStringArray("rest_actions_filters", null));
+            } else if (request.hasParam("rest_actions_filters")) {
```
Comment (mgodwan) [inline server/src/main/java/org/opensearch/rest/action/admin/cluster/RestNodesStatsAction.java:200]: Do we need validations on the provided values for `rest_actions_filters` as well?

**Review by mgodwan — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodesStatsRequest.java (hunk 1 above)
Comment (vzez) [reply server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodesStatsRequest.java:84]: if filters not provided, we'll get status count for all rest rest actions having count > 0

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/RestActionsStats.java (hunk 2 above)
Comment (vzez) [reply server/src/main/java/org/opensearch/action/admin/cluster/node/stats/RestActionsStats.java:29]: I think status count won't go beyond INT range, so making it LONG, will cause memory wastage.

File: server/src/main/java/org/opensearch/action/admin/cluster/node/stats/RestActionsStats.java (hunk 3 above)
Comment (vzez) [reply server/src/main/java/org/opensearch/action/admin/cluster/node/stats/RestActionsStats.java:30]: 1. What sort of doc is req here? could you please explain.
2. For Map @line30 , yes, the status code is the Key. For Map @line29 "action name" is the key.

File: server/src/main/java/org/opensearch/rest/RestActionsService.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * A service to fetch rest actions responses count
+ *
+ * @opensearch.internal
+ */
+public class RestActionsService {
```
[the remark this answers is not in the document]
Comment (vzez) [reply server/src/main/java/org/opensearch/rest/RestActionsService.java:26]: Indeed, "RestActionsStatsService" better suits here and gives a clear info about class as well.
I came up with another one also -"RestActionsStatsCountService" , will it be ok?

File: server/src/main/java/org/opensearch/rest/RestActionsService.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    private Map<Integer, Integer> getRestActionsStatusTotalCount(Map<String, Map<Integer, Integer>> mapHandlerStatusCount) {
+        final Map<Integer, Integer> totalStatusCount = new TreeMap<>();
+        mapHandlerStatusCount.entrySet().stream().flatMap(entry -> entry.getValue().entrySet().stream()).forEach(entry -> {
+            Integer count = totalStatusCount.get(entry.getKey());
+            if (null == count) count = 0;
+            count += entry.getValue();
+            totalStatusCount.put(entry.getKey(), count);
+        });
+        return totalStatusCount;
+    }
```
[the remark this answers is not in the document]
Comment (vzez) [reply server/src/main/java/org/opensearch/rest/RestActionsService.java:48]: I'll cross check, if not added, I'll add them.

File: server/src/main/java/org/opensearch/rest/RestActionsService.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    public void addRestHandler(BaseRestHandler handler) {
+        Objects.requireNonNull(handler);
+        if (handler.getName() == null) {
+            throw new IllegalArgumentException("handler of type [" + handler.getClass().getName() + "] does not have a name");
+        }
+        handlers.putIfAbsent(handler.getName(), new TreeMap<>());
```
[the remark this answers is not in the document]
Comment (vzez) [reply server/src/main/java/org/opensearch/rest/RestActionsService.java:55]: No, putIfAbsent only puts value if no entry exists for the key and returns null, if key is already there it doesn't replace exiting value but returns existing value.
ex: 
map.putIfAbsent("exist", "no"); --- returns null
map.putIfAbsent("exist", "yes"); --- returns "no" 
there is no error or exception

File: server/src/main/java/org/opensearch/rest/RestController.java [hunk 8]
```diff
@@ -572,6 +588,10 @@ public boolean detailedErrorsEnabled() {
         public void sendResponse(RestResponse response) {
             close();
             delegate.sendResponse(response);
+            if (restHandler instanceof BaseRestHandler) restActionsService.putStatus(
+                ((BaseRestHandler) restHandler).getName(),
+                response.status()
+            );
```
[the remark this answers is not in the document]
Comment (vzez) [reply server/src/main/java/org/opensearch/rest/RestController.java:594]: I already formatted code using ./gradlew spotlessApply. 
I guess had the code not formatted using above command, it couldn't be pushed.

File: server/src/main/java/org/opensearch/rest/action/admin/cluster/RestNodesStatsAction.java (hunk 4 above)
Comment (vzez) [reply server/src/main/java/org/opensearch/rest/action/admin/cluster/RestNodesStatsAction.java:200]: I did have thoughts about it while coding but the thing is we need to maintain the list of all rest actions and whenever a new rest action would be introduced, it must be added to that list. So I think it's un-necessary. As per current implementation if filter contain invalid action name it won't be returned as response.

Or we have to use RestActionsService (class in which all handlers are being registered) ref here, again it creates un-necessary dependency which I think is not a good design.

File: server/src/main/java/org/opensearch/rest/RestActionsService.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        if (handlers.containsKey(handlerName)) {
+            increment(handlers.get(handlerName), status.getStatus());
+        }
+    }
+
+    private void increment(Map<Integer, Integer> statusCount, int status) {
+        Integer count = statusCount.get(status);
+        if (null == count) count = 0;
+        statusCount.put(status, ++count);
+    }
```
[the remark this answers is not in the document]
Comment (vzez) [reply server/src/main/java/org/opensearch/rest/RestActionsService.java:68]: Rather using AtomicLong , we can make the Map thread-safe?

**Comment by vzez** [comment]
Closing this PR , new PR raised by Prabhat #5334

**Pull request closed without merging**